_docs/cardano/2017-01-07-addresses.md

```diff
@@ -122,25 +85,71 @@ To quote Bitcoin Wiki,
 > these bitcoins, or a password might be required, or the requirements could be
 > completely unique.
 
-`ScriptHash` addresses are constructed as follows:
-
-    address' ← 0x01 + address_hash(serialize(validator_script))
-    address ← toBase58(address' + crc32(address'))
-
 ## Redeem Addresses
 
 Redeem addresses are Pay To PubKey Hash (P2PKH). Such an address contains the hash
 of redeem public key, and this key is actually [Ed25519](http://ed25519.cr.yp.to/)
 public key.
 
-Redeem addresses are constructed as follows:
- 
-    address' ← 0x02 + address_hash(redeem_public_key)
-    address ← toBase58(address' + crc32(address'))
-
 ## Other Address Types
 
-In the future, we may use the update system to introduce other address types
-with different values in the `type` field. [See
-more](/cardano/update-mechanism/#soft-fork-updates) on extending the system in
+In the future, we may use the update system to introduce other address types. Please
+[see more](/cardano/update-mechanism/#soft-fork-updates) on extending the system in
 non-breaking fashion.
+
+## Address Structure
+
+Address consists of 3 parts:
+
+*  address root,
+*  address attributes,
+*  address type.
+
+We can imagine an address as a JSON-like structure, for example:
+
+```
+Address {
+    addrRoot = AbstractHash e63175c654dfd93a9290342a067158dc0f57a1108ddbd8cace3839bd,
+    addrAttributes = Attributes {
+        data: AddrAttributes {
+            aaPkDerivationPath = Nothing,
+            aaStakeDistribution = BootstrapEraDistr
+        } 
+    },
+    addrType = ATPubKey
+}
+```
+
+`addrRoot` is the BLAKE2b-224 hash of the tuple made from `addrType`, `addrSpendingData` and `addrAttributes`.
+
+`addrSpendingData is a special value which is bound to an address and must be revealed in order to spend coins belonging to
+this address. For example, for public key address this value contains the public key. In this case it is impossible to change
+address attributes without knowing of the public key because if the attributes have been changed the whole address becomes
+invalid.
+
+`addrAttributes` include important attributes of each address: derivation path and stake distribution.
+
+For more info about derivation path please read [HD Wallets in Cardano SL](https://cardanodocs.com/technical/hd-wallets/) chapter.
+
+For more info about stake distribution please read [Transactions in Cardano SL](https://cardanodocs.com/cardano/transactions/#stake-distribution)
+chapter.
+
+Value of `addrType` corresponds to address type as was mentioned above, in this example it is a public key address.
+
+### Length
+
+Addresses may have different lengths depending on address type and additional data in it.
+
+For example, this address
+
+```
+Ae2tdPwUPEZKmwoy3AU3cXb5Chnasj6mvVNxV1H11997q3VW5ihbSfQwGpm
+```
+
+and this one
+
+```
+4swhHtxKapQbj3TZEipgtp7NQzcRWDYqCxXYoPQWjGyHmhxS1w1TjUEszCQT1sQucGwmPQMYdv1FYs3d51KgoubviPBf
+```
+
+are both public key addresses, but have different lengths.
```

Comment:
At this point it's obvious for me that they have different lengths :smile: 
Either remove last part of the sentence, or say why they differ in length in this particular case

Reply:
:-) Yeah, you're right!